airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/DestinationTask.kt

```diff
@@ -0,0 +1,411 @@
+package io.airbyte.cdk.write
+
+import java.util.concurrent.atomic.AtomicInteger
+import kotlinx.coroutines.delay
+
+data class Concurrency(
+    val scopeId: String,
+    val perSync: Int = 0,
+    val perStream: Int = 0,
+    val exclusions: Set<String> = emptySet()
+)
+
+/**
+ * A task or block of tasks that returns a value.
+ *
+ * T is the value returned by executing all underlying tasks
+ * U is the value returned by running the individual task
+ *
+ * For Control Tasks
+ *   * T is the "underlying value" (returned by a nested DestinationTask)
+ *   * U is a list of Task(s) that return T (or return tasks that return T)
+ * For DestinationTasks
+ *   * U==T (ie, the "underlying value" is the same as the value returned)
+ */
+abstract class AbstractDestinationTask<T, out U> {
+    /**
+     * A task that when executed yields a task generated
+     * from the underlying value of another task. The client
+     * will never need to create or extend this directly.
+     */
+    abstract class FromValue<T, U>: ControlTask<U, AbstractDestinationTask<U, *>>()
+
+    /**
+     * A task that when executed yields a task generated
+     * from the underlying task itself. The client will never
+     * need to create or extend this directly.
+     */
+    abstract class FromTask<T, U>: ControlTask<U, AbstractDestinationTask<U, *>>()
+
+    /**
+     * Execute the task and return the value.
+     */
+    abstract suspend fun execute(): U
+
+    /**
+     * Given a function from this task's underlying value to another task,
+     * yield a task that when executed (or when all underlying tasks are executed),
+     * returns a list of tasks generated from the returned value(s).
+     */
+    abstract fun <V, W : AbstractDestinationTask<V, *>> mapValue(taskFor: (T) -> W): FromValue<T, V>
+
+    /**
+     * Given a function from this task to another task,
+     * yield a task that when executed (or when all underlying tasks are executed),
+     * returns a list of tasks generated from the underlying task.
+     */
+    abstract fun <V, W : AbstractDestinationTask<V, *>> mapTask(taskFor: (AbstractDestinationTask<T, *>) -> W): FromTask<T, V>
```

Comment:
In my experience, `fun interface` are nicer than lambda arguments if only because they age better. For example, what if at some point you also want to make your argument closable? It's a lot easier to do when you're already passing an interface. Instantiating a `fun interface` is almost as lightweight as a lambda so there's not much of a downside.